src/uct/ib/ud/verbs/ud_verbs.c

```diff
@@ -393,6 +395,20 @@ uct_ud_verbs_iface_poll_tx(uct_ud_verbs_iface_t *iface, int is_async)
     return 1;
 }
 
+static UCS_F_ALWAYS_INLINE size_t uct_ud_verbs_iface_get_gid_len(void *packet)
+{
+  /* The GRH will contain either an IPv4 or IPv6 header. If the former is
+   * present the header will start at offset 20 in the buffer otherwise it
+   * will start at offset 0. Since the two headers are of fixed size (20 or
+   * 40 bytes) this means we will either see 0x6? at offset 0 (IPv6) or 0x45
+   * at offset 20. The detection is a little tricky for IPv6 given that the
+   * first 20B are undefined for IPv4. To overcome this the first byte of
+   * the posted receive buffer is set to 0xff.
+   */
+  return ((((unsigned char *)packet)[0] & 0xf0) == 0x60) ? UCS_IPV6_ADDR_LEN :
+        (((unsigned char *)packet)[20] == 0x45) ? UCS_IPV4_ADDR_LEN : 0;
```

Comment:
do we really have to check also byte [20]?

Reply:
Intended to catch an invalid grh. Figured it is worth warning if the packet neither starts with 0x6? or has 0x45 at offset 20.

Reply:
IMO better remove for optimization

Reply:
Fair enough. Removing in next push.